# Face & eye Detection in HAAR Cascade Classifiers

## Import packages

In [1]:
import os

In [2]:
pip install opencv-python

Note: you may need to restart the kernel to use updated packages.


In [3]:
import cv2
import numpy as np

## List of Directory items

In [4]:
os.listdir()

['.ipynb_checkpoints',
 'Face_And_Eye_detection_Opencv_CNN.ipynb',
 'haarcascades',
 'images']

## Load the FrontalFace Cascade classifier

In [5]:
face_classifier = cv2.CascadeClassifier('haarcascades/haarcascade_frontalface_default.xml')

## Load the image then convert into grayscale

In [6]:
image = cv2.imread('images/trump-jp.jpg')

In [7]:
a = cv2.imshow('image1',image)
a
cv2.waitKey(0)

-1

## Detect the MultiScale of the image

In [8]:
faces = face_classifier.detectMultiScale(image,1.3,4)
faces

array([[291, 113, 283, 283]])

    [291, 113, 283, 283]
    [ x1 , y1 , x2 , y2]
    [ x  , y  , w  , h ] As per the python program.
    
    By the (x1,y1) &(x2,y2) we will find the another two rectangle points. Like,
    (x1,y1)             (x1+x2,y1+y2)
       .----------------------.
       |                      |
       |                      |
       |     Image Face       |
       |                      |
       |                      |
       |                      |
       .----------------------.
    (x2,y2)             (x1+x2,y1+y2)

In [9]:
## When No face detected , face_classifier return and empty tuple
if faces is ():
    print('No faces found')

<>:2: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:2: SyntaxWarning: "is" with a literal. Did you mean "=="?
<ipython-input-9-63a77ade5770>:2: SyntaxWarning: "is" with a literal. Did you mean "=="?
  if faces is ():


## Draw the rectangle
### Face Detection

In [10]:
# we iterate through faces array and draw and rectangle.

# Detect the Face
for (x,y,w,h) in faces:
    cv2.rectangle(image,(x,y), (x+w,y+h),(100,0,121),10)
    cv2.imshow('Trump_Face',image)
    cv2.waitKey(0)
cv2.destroyAllWindows()

Now we did the face detection. 
After this we will try the eye detection.

### Eye Detection

In [11]:
face_classifier = cv2.CascadeClassifier('haarcascades/haarcascade_frontalface_default.xml')
eye_classfier = cv2.CascadeClassifier('haarcascades/haarcascade_eye.xml')

In [12]:
image = cv2.imread('images/trump-jp.jpg')
gray = cv2.cvtColor(image,cv2.COLOR_BGR2GRAY)

<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<ipython-input-14-aec39272f535>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
  if faces is ():


In [18]:
faces = face_classifier.detectMultiScale(gray)

if faces is ():
    print('No Face found')
    
for (x,y,w,h) in faces:
    cv2.rectangle(image,(x,y), (x+w,y+h),(127,0,255),2)
    cv2.imshow('Trump_Face',image)
    roi_gray = gray[y:y+h,x:x+w]
    roi_color = image[y:y+h,x:x+w]

eyes = eye_classfier.detectMultiScale(roi_gray,1.5,1)

if eyes is ():
    print('No Eyes found')

for (ex,ey,ew,eh) in eyes:
    cv2.rectangle(roi_color,(ex,ey), (ex+ew,ey+eh),(225,225,0),2)
    cv2.imshow('Trump_Face',image)

    
cv2.waitKey(0)
cv2.destroyAllWindows()

<>:3: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:14: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:3: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:14: SyntaxWarning: "is" with a literal. Did you mean "=="?
<ipython-input-18-9a73a6478591>:3: SyntaxWarning: "is" with a literal. Did you mean "=="?
  if faces is ():
<ipython-input-18-9a73a6478591>:14: SyntaxWarning: "is" with a literal. Did you mean "=="?
  if eyes is ():


No Eyes found


In [19]:
eyes

()